### PR TITLE
[Federation] Disallow extensions without base definitions

### PR DESCRIPTION
**Pull request by JakeDawkins**
## What is this

Previously, when building federated services, composition allowed you to `extend` a type without ever defining it in another service. This PR removes that :) 

## What was done

There were a few changes that needed to be made to make this work.

1. remove the block of code in `compose.ts` that enabled this behavior to begin with.
2. Add empty `Query` and optional `Mutation` definitions to the `definitionsMap`. This is needed because the old block used to add the query and mutation definitions for us. Query and Mutation definitions found in a service are always treated as _extensions_ to make sure they never have an `owner`.
3. I added a validation rule to make sure this is properly erroring. It checks for 2 things: 
    1. To make sure there is a definition for any extension
    2. make sure the extension is of the same `Kind` (no `type`s extending `interface`s or any of that nonsense
    3. these things only work with types and interfaces at the moment. It's easy to add more though.

## Misc

1. Fixed an integration test that previously left off the `books` service, leading to a validation error because there wasn't a `Book` definition :)
2. Added a test at the `composeAndValidate` level for this, just to make sure nothing was getting swallowed.

### CODE REVIEW

File: packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+          'EXTENSION_WITH_NO_BASE',
+          logServiceAndType(serviceName, typeName) +
+            `\`${typeName}\` is an extension type, but \`${typeName}\` is not defined in any service`,
+        ),
+      );
+    }
+  };
+
+  return {
+    ObjectTypeExtension: checkExtension,
```
Comment (JakeDawkins) [inline packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts]: do we want to support this for any other types?

File: packages/apollo-federation/src/composition/compose.ts [hunk 2]
```diff
@@ -220,35 +220,26 @@ export function buildMapsFromServiceList(serviceList: ServiceDefinition[]) {
...
+  const emptyDefinition = (name: string) => ({
+    kind: Kind.OBJECT_TYPE_DEFINITION,
+    name: { kind: Kind.NAME, value: name },
+    fields: [],
+    serviceName: null,
+  });
+  if (!definitionsMap['Query'])
+    definitionsMap['Query'] = [emptyDefinition('Query')];
+  if (extensionsMap['Mutation'] && !definitionsMap['Mutation'])
+    definitionsMap['Mutation'] = [emptyDefinition('Mutation')];
```
Comment (JakeDawkins) [inline packages/apollo-federation/src/composition/compose.ts]: I tried adding these empty definitions to the schema constructor, but it errored because they required fields

File: packages/apollo-federation/src/composition/compose.ts [hunk 3]
```diff
@@ -220,35 +220,26 @@ export function buildMapsFromServiceList(serviceList: ServiceDefinition[]) {
...
-  }
+  // We need to have a base definition of Query and potentitally Mutation to
+  // extend later on...
+  const emptyDefinition = (name: string) => ({
+    kind: Kind.OBJECT_TYPE_DEFINITION,
+    name: { kind: Kind.NAME, value: name },
+    fields: [],
+    serviceName: null,
+  });
+  if (!definitionsMap['Query'])
```
Comment (jbaxleyiii) [inline packages/apollo-federation/src/composition/compose.ts]: This is a nit, but why not use `definitionsMap.Query`?

File: packages/apollo-federation/src/composition/compose.ts (hunk 2 above)
Comment (jbaxleyiii) [reply packages/apollo-federation/src/composition/compose.ts]: Does this get around that? Seems like an odd working around the type system?

File: packages/apollo-federation/src/composition/compose.ts [hunk 4]
```diff
@@ -315,6 +306,7 @@ export function addFederationMetadataToSchemaNodes({
     { serviceName: baseServiceName, extensionFieldsToOwningServiceMap },
   ] of Object.entries(typeToServiceMap)) {
     const namedType = schema.getType(typeName) as GraphQLNamedType;
+    if (!namedType) continue;
```
Comment (jbaxleyiii) [inline packages/apollo-federation/src/composition/compose.ts]: Why was this added? (can we add a comment?)

File: packages/apollo-federation/src/composition/validate/sdl/__tests__/possibleTypeExtensions.test.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+      serviceList,
+    );
+    const errors = validateSDL(definitions, schema, [PossibleTypeExtensions]);
+    schema = extendSchema(schema, definitions, { assumeValidSDL: true });
+    errors.push(...validateSDL(extensions, schema, [PossibleTypeExtensions]));
+    expect(errors).toMatchInlineSnapshot(`
+            Array [
+              Object {
+                "code": "EXTENSION_OF_WRONG_KIND",
+                "message": "[serviceA] Product -> \`Product\` was originally defined as a InputObjectTypeDefinition and can only be extended by a InputObjectTypeExtension. serviceA defines Product as a ObjectTypeExtension",
```
Comment (jbaxleyiii) [inline packages/apollo-federation/src/composition/validate/sdl/__tests__/possibleTypeExtensions.test.ts]: YAS

File: packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    const serviceName = node.serviceName;
+    if (!serviceName) return;
+
+    if (defNode) {
+      const expectedKind = defKindToExtKind[defNode.kind];
+      const baseKind = defNode.kind;
+      if (expectedKind !== node.kind) {
+        context.reportError(
+          errorWithCode(
+            'EXTENSION_OF_WRONG_KIND',
```
Comment (jbaxleyiii) [inline packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts]: This is so helpful

File: packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts (hunk 1 above)
Comment (jbaxleyiii) [reply packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts]: Not currently! In the future Unions may come into play, but we have a lot of thinking on that first 😆

File: packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+const defKindToExtKind: { [kind: string]: string } = {
+  [Kind.SCALAR_TYPE_DEFINITION]: Kind.SCALAR_TYPE_EXTENSION,
+  [Kind.OBJECT_TYPE_DEFINITION]: Kind.OBJECT_TYPE_EXTENSION,
+  [Kind.INTERFACE_TYPE_DEFINITION]: Kind.INTERFACE_TYPE_EXTENSION,
+  [Kind.UNION_TYPE_DEFINITION]: Kind.UNION_TYPE_EXTENSION,
+  [Kind.ENUM_TYPE_DEFINITION]: Kind.ENUM_TYPE_EXTENSION,
+  [Kind.INPUT_OBJECT_TYPE_DEFINITION]: Kind.INPUT_OBJECT_TYPE_EXTENSION,
+};
+
+function typeToKind(type: any) {
```
Comment (jbaxleyiii) [inline packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts]: Are these not already part of graphql-js

File: packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    }
+  };
+
+  return {
+    ObjectTypeExtension: checkExtension,
+    InterfaceTypeExtension: checkExtension,
+  };
+}
+
+function typeToExtKind(type: any) {
```
Comment (jbaxleyiii) [inline packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts]: Are these not already part of graphql-js

Comment (jbaxleyiii) [reply packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts]: Also, can we make the typings more specific here?

File: packages/apollo-federation/src/composition/compose.ts (hunk 4 above)
Comment (JakeDawkins) [reply packages/apollo-federation/src/composition/compose.ts]: todo: fix typecasting/comment/typeToServiceMap

**Comment by JakeDawkins** [comment]
Had a fun like 15 files of linting errors, so this PR also fixes those 😞
